NebulaClient/MonoBehaviours/MultiplayerSession.cs

```diff
@@ -114,5 +138,107 @@ private void OnVegeMined(VegeMined packet)
 	    {
             GameMain.localPlanet?.factory?.RemoveVegeWithComponents(packet.VegeID);
 	    }
+
+        private void OnHandshakeResponse(HandshakeResponse packet)
+        {
+            if(packet.IsFirstPlayer)
+            {
+                // We want to host the world here
+                GameDesc gameDesc = new GameDesc();
+                gameDesc.SetForNewGame(UniverseGen.algoVersion, 1, 64, 1, 1f);
+                DSPGame.StartGameSkipPrologue(gameDesc);
+                PlayerManager.WeAreMainPlayer = true;
+                statusBox.FadeOut();
+            }
+            else
+            {
+                statusBox.FadeOut();
+                statusBox = null;
+                statusBox = UIMessageBox.Show("Loading", "Loading state from server, please wait", null, UIMessageBox.INFO);
+
+                foreach(var playerId in packet.OtherPlayerIds)
+                {
+                    PlayerManager.AddRemotePlayer(playerId);
+                }
+            }
+        }
+
+        private void OnInitialState(InitialState packet)
+        {
+            StartCoroutine(InitialStateSyncUtils.DownloadInitialState(packet.URI, statusBox, () => 
+            {
+                Debug.Log("Loading from: " + GameConfig.gameSaveFolder + "INITIALMPSTATE.dsv");
+                DSPGame.StartGame("INITIALMPSTATE");
+                statusBox.FadeOut();
+                statusBox = null;
+
+                Client.SendPacket(new SyncComplete());
+            }));
+        }
+
+
+        private void OnSyncComplete(SyncComplete packet)
+        {
+            GameMain.Resume();
+            if(statusBox != null)
+            {
+                statusBox.FadeOut();
+                statusBox = null;
+            }
+        }
+
+        private class InitialStateSyncUtils
```

Comment:
I would prefer to have this class in its own file for ex: "NebulaClient/Utils/InitialStateSyncUtils.cs"